### DISCUSSION TITLE
Troubles with repeating delimiter

### DISCUSSION THREAD

**Discussion by ctron** [eclipse-langium/langium#965] [Q&A] [answered]
I am trying to get a comma separated list going, where a trailing comma is allowed:

```
pattern bar = {
    field1: foo,
    field2: boolean,
    field3: {},  <1>
}
```
<1> The final command may be present or not

I am using the playground for experimenting with this. Here is the link to my original version: [playground link](https://langium.org/playground?grammar=OYJwhgthYgBAIge2NAUKgpgOwC4gJ6wCyiAJhgDYBcqsdsAFAA5g44YhYDOA1ALwAFVu04BKAFQBudELYcsNerADkLOZ2WwskDHwCS8FX0058fACr4mGaakvXFsAEoYAZh2wBjDLAA%2BsAHkAIwArDE8cP1gBEABLCDJYsApbF3cQLwxHUz4AbVkRBQMAXVsY%2BMTkxwBlPFisYCiAIURECgwwLFtauIbHOgBvevTYHvrgAF8VLjqG5VsWto6FWnohrBHF9s6p5SDW7ax59GCwiP6VAc0GV1jKUi5YflgAMTuKUhUAGmUJRlv7o9nm97qIAPwqCbKVZ0Wwgj4XbQQXQGRiIJg4WKIbQUMHGMG-CHKKgmMz2GzoAAWsVI5CwsBEEHqyVgAHVqlRYAB6AA6vC50kZzIosAMnK5uQA%2BmAALQALwAgjKAFrFXI8gDukuK4gFVJpdIZHCZOOIABlJQBhAJEIgAUQAcuZxTzeeJ1VwedUdWCeeJXXrqbTsEaQCaWdULdbbY7ndzXa7cgA9HlYHkgHV6oA&content=A4QwLmCmBOB2AEAzA9s%2BBeeBvAUPfSAlpADYAmAXPAM5jSGwDmOAvjjqBDAgEYjQZseAomLkAjFRTIANMPyjSZAExUeqEpBCw5BIkoDMVLCzlsO4KHHh8AXoJM4gA)

This results in the problem, that it always seems to require a comma, my expectation would be, that if the first section (`fields += Field ','`) no longer matches, the next one is processes, which doesn't require a comma. However, that doesn't seem to work.

If I change the rule of `Object`  to be a bit more explicit:

```
Object:
    '{' ( (fields += Field) (',' fields += Field)* | (fields += Field ',')+ )'}'
    ;
```

My expectation would be that it tries either branch, on of them succeeding. However, what I get is a frozen playground, and the following error in the browser console:

```
Request initialize failed with message: Parser Definition Errors detected:
 Ambiguous Alternatives Detected: <1 ,2> in <OR1> inside <Object> Rule,
<ID, ?:KW, ::KW> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.
-------------------------------
Ambiguous Alternatives Detected: <1 ,2> in <OR1> inside <Object> Rule,
<ID, ::KW, ID> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.
-------------------------------
Ambiguous Alternatives Detected: <1 ,2> in <OR1> inside <Object> Rule,
<ID, ::KW, {:KW> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.
-------------------------------
Ambiguous Alternatives Detected: <1 ,2> in <OR1> inside <Object> Rule,
<ID, ::KW, string:KW> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.
-------------------------------
Ambiguous Alternatives Detected: <1 ,2> in <OR1> inside <Object> Rule,
<ID, ::KW, boolean:KW> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.
```

Playground link [here](https://langium.org/playground?grammar=OYJwhgthYgBAIge2NAUKgpgOwC4gJ6wCyiAJhgDYBcqsdsAFAA5g44YhYDOA1ALwAFVu04BKAFQBudELYcsNerADkLOZ2WwskDHwCS8FX0058fACr4mGaakvXFsAEoYAZh2wBjDLAA%2BsAHkAIwArDE8cP1gBEABLCDJYsApbF3cQLwxHUz4AbVkRBQMAXVsY%2BMTkxwBlPFisYCiAIURECgwwLFtauIbYevSuaLiE0iTqWnoAbwGOWB764ABfFS46huVbFraOrH6sQeGKsarJuhmDue32zpXlINabrE30YLCIxzplKc0GRldYpRSEN%2BLAAGKAiikUSMZQAGk0AKBIL44Mh0PEUQYSKhKLRQJUCNEPFgomUS2UZ1gtghQM%2BWh0%2BkMDEQTBwsUQ2goAH5jNzlKJuSoqCYzPYbOgABaxUjkPYiCD1ZKwADq1SosAA9AAdXia6QKpUUWAGDWa3IAfTAAFoAF4AQWtAC1irltQB3C3FcT6qUyuWwQ1c4gAGQtAGEAkQiABRABy5jN2p14jdXG11W93O14mTvulsuwgY4iuD1TDkej8cTWuTydyAD1tVhtSBvb6gA&content=A4QwLmCmBOB2AEAzA9s%2BBeeBvAUPfSAlpADYAmAXPAM5jSGwDmOAvjjqBDAgEYjQZseAomLkAjFRTIANMPyjSZAExUeqEpBCw5BIkoDMVLCzlsO4KHHh8AXoJM4gA)

My expectation would be, if there are errors like this, they get shown to the user and modifying the grammar into an "ok" state, would revive the editor.

**Comment by msujew** [comment, marked as answer]
Hey @ctron,

as for the playground issue: We already have [an issue on the langium-website](https://github.com/langium/langium-website/issues/130) for this.

Regarding your grammar: `( (fields += Field) (',' fields += Field)* | (fields += Field ',')+ )` is ambiguous, since both alternatives start with the `Field` subrule and the static lookahead isn't long enough to figure out how to disambiguate it. There's an easier way to accomplish your goal:

```
Object:
    '{' ( (fields += Field) (',' (fields += Field)?)* )? '}'
;
```

**Reply by ctron** [reply]
Awesome thanks! So I do understand that the second version didn't work. But shouldn't the first one have worked too:

```
Object:
    '{' (fields += Field ',')* (fields += Field)? '}'
    ;
```

**Reply by msujew** [reply]
Well, no, and it's due to our parsers lookahead algorithm. Basically it doesn't really decide whether to parse `(fields += Field ',')*` OR `(fields += Field)?`, but instead it tries to parse `(fields += Field ',')*` eagerly due to it's limited lookahead. The parser doesn't look ahead enough to figure out that `,` is the disambiguating character. While this makes grammar construction a bit less intuitive it has a dramatic impact on parsing performance. Effectively reducing the cost of parsing from O(n³) to O(n).

You should generally avoid constructs that have multiple optional groups right after another that start with the same tokens.
